pkg/controllers/cluster/taint_manager.go

```diff
@@ -0,0 +1,347 @@
+		Selector: fields.OneTermEqualSelector(crbClusterKeyIndex, cluster.Name),
+	}); err != nil {
+		klog.ErrorS(err, "Failed to list ClusterResourceBindings", "cluster", cluster.Name)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	for i := range crbList.Items {
+		key, err := keys.FederatedKeyFunc(cluster.Name, &crbList.Items[i])
+		if err != nil {
+			klog.Warningf("Failed to generate key for obj: %s", crbList.Items[i].GetObjectKind().GroupVersionKind())
+			continue
+		}
+		tc.clusterBindingEvictionWorker.Add(key)
+	}
+	return controllerruntime.Result{RequeueAfter: tc.ClusterTaintEvictionRetryFrequency}, nil
+}
+
+// Start starts an asynchronous loop that handle evictions.
+func (tc *NoExecuteTaintManager) Start(ctx context.Context) error {
+	bindingEvictionWorkerOptions := util.Options{
+		Name:          "binding-eviction",
+		KeyFunc:       nil,
+		ReconcileFunc: tc.syncBindingEviction,
+	}
+	tc.bindingEvictionWorker = util.NewAsyncWorker(bindingEvictionWorkerOptions)
+	tc.bindingEvictionWorker.Run(tc.ConcurrentReconciles, ctx.Done())
+
+	clusterBindingEvictionWorkerOptions := util.Options{
+		Name:          "cluster-binding-eviction",
+		KeyFunc:       nil,
+		ReconcileFunc: tc.syncClusterBindingEviction,
+	}
+	tc.clusterBindingEvictionWorker = util.NewAsyncWorker(clusterBindingEvictionWorkerOptions)
+	tc.clusterBindingEvictionWorker.Run(tc.ConcurrentReconciles, ctx.Done())
+
+	<-ctx.Done()
+	return nil
+}
+
+func (tc *NoExecuteTaintManager) syncBindingEviction(key util.QueueKey) error {
+	fedKey, ok := key.(keys.FederatedKey)
+	if !ok {
+		klog.Errorf("Failed to sync binding eviction as invalid key: %v", key)
+		return fmt.Errorf("invalid key")
+	}
+	cluster := fedKey.Cluster
+	keyStr := fedKey.String()
+
+	binding := &workv1alpha2.ResourceBinding{}
+	if err := tc.Client.Get(context.TODO(), types.NamespacedName{Namespace: fedKey.Namespace, Name: fedKey.Name}, binding); err != nil {
+		// The resource no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get binding %s: %v", fedKey.NamespaceKey(), err)
+	}
+
+	if !binding.DeletionTimestamp.IsZero() || !util.IsTargetClusterExist(binding.Spec.Clusters, cluster) {
+		return nil
+	}
+
+	needEviction, tolerationTime, err := tc.needEviction(cluster, binding.Annotations)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if binding needs eviction", "binding", klog.KObj(binding))
+		return err
+	}
+
+	// Case 1: Need eviction now.
+	// Case 2: need eviction after toleration time. If time is up, do eviction right now.
+	// Case 3: Tolerate forever, we do nothing.
+	evictNow := false
+	if needEviction || tolerationTime == 0 {
+		evictNow = true
+	} else if tolerationTime > 0 {
+		now := time.Now()
+		if startTime, exist := tc.evictionStartTimeMap.get(keyStr); exist {
+			if startTime.Add(tolerationTime).Before(now) {
+				evictNow = true
+			}
+		} else {
+			tc.evictionStartTimeMap.set(keyStr, now)
```

Comment:
I think we should check if the key exists, if not then set the start time

Reply:
Indeed, the check is done first. 
```go
		if startTime, exist := tc.evictionStartTimeMap.get(keyStr); exist {
			...
		} else {
			tc.evictionStartTimeMap.set(keyStr, now)
			...
		}
```

Reply:
Oh yes I missed it